### PR TITLE
Json collector

### PR DESCRIPTION
**Pull request by odain-cbd**
review with Anne-Catherine

### CODE REVIEW

File: test/jsoncollector_test.inc.php [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                }
+            }
+        }
+    }
+    /*
+     * @param string $initDir initial directory of file params.distrib.xml
+     */
+    private function replaceTranslateRelativePathInParam($initDir)
+    {
+        $sContent = str_replace("APPROOT", APPROOT, file_get_contents(APPROOT . $initDir."/params.distrib.xml"));
```
Comment (Hipska) [inline test/jsoncollector_test.inc.php:76]: Could this also be relevant for outside the tests?

**Review by Hipska — CHANGES_REQUESTED** [review]

File: core/testconnection.inc.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+define('APPROOT', dirname(__FILE__).'/../');
+require_once(APPROOT.'core/parameters.class.inc.php');
+require_once(APPROOT.'core/utils.class.inc.php');
+require_once(APPROOT.'core/restclient.class.inc.php');
+
+if (DIRECTORY_SEPARATOR === '\\') {
+    // windows
+    print 'Windows not supported!'.PHP_EOL;
+    exit(1);
+}
```
[the remark this answers is not in the document]
Comment (Molkobain) [reply core/testconnection.inc.php:12]: @Hipska has a point

[the remark this answers is not in the document]
Comment (odain-cbd) [reply core/testconnection.inc.php:12]: i agree.

FYI at the very beginning that script contained unix specific commands (curl/wget) to get network info and easily troubleshoot why itop API is not reachable. but i removed that part.

[the remark this answers is not in the document]
Comment (odain-cbd) [reply core/testconnection.inc.php:12]: I will fix that. but from my point of view i would never support windows....  :)

[the remark this answers is not in the document]
Comment (Molkobain) [reply core/testconnection.inc.php]: Keep in mind that many of our customers and people in the community are using IIS as a web server, so we should always maintain compatibility.

[the remark this answers is not in the document]
Comment (Hipska) [reply core/testconnection.inc.php]: And the test is still based on `DIRECTORY_SEPARATOR` Which was the main reason for my request to change.

File: core/testconnection.inc.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+define('APPROOT', dirname(__FILE__).'/../');
+require_once(APPROOT.'core/parameters.class.inc.php');
+require_once(APPROOT.'core/utils.class.inc.php');
+require_once(APPROOT.'core/restclient.class.inc.php');
+
+if (DIRECTORY_SEPARATOR === '\\') {
```
Comment (Hipska) [inline core/testconnection.inc.php:8]: This test can be better

**Review by Hipska — CHANGES_REQUESTED** [review]

File: core/testconnection.inc.php (hunk 3 above)
Comment (odain-cbd) [reply core/testconnection.inc.php]: test and log are useless. i removed them

File: core/testconnection.inc.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (odain-cbd) [reply core/testconnection.inc.php]: test removed

File: core/jsoncollector.class.inc.php [hunk 4]
```diff
@@ -327,6 +320,15 @@ public function Fetch()
                     throw new Exception("Missing columns in the Json file.");
                 }
             }
+            //check if all expected fields are in array. If not add it with null value
+            foreach($this->aCSVHeaders as $sHeader)
+            {
+               if(!isset($aDataToSynchronize[$sHeader]))
+               {
+                   $aDataToSynchronize[$sHeader]="";
```
Comment (Hipska) [inline core/jsoncollector.class.inc.php:328]: This is not a null value:
```suggestion
                   $aDataToSynchronize[$sHeader]=null;
```

File: core/jsoncollector.class.inc.php [hunk 5]
```diff
@@ -294,18 +290,15 @@ public function Fetch()
...
-                    Utils::Log(LOG_DEBUG, "this->aJson[$sTag]: " . json_encode($aValue));
-
-                    if (!$bFind) {
-                        Utils::Log(LOG_ERR, "[" . get_class($this) . "] Failed to find path '$sPath' until data in json file: '$this->sURL'.");
-                        return false;
-                    }
                 }
-                $aDataToSynchronize[$key] = $aValue;
+                if ($bFind) {
+                    Utils::Log(LOG_DEBUG, "aDataToSynchronize[$sTag]: " . json_encode($aValue));
```
Comment (Hipska) [inline core/jsoncollector.class.inc.php:295]:
```suggestion
                    Utils::Log(LOG_DEBUG, "aDataToSynchronize[$key]: " . json_encode($aValue));
```

**Review by Hipska — CHANGES_REQUESTED** [review]

File: core/csvcollector.class.inc.php [hunk 6]
```diff
@@ -28,258 +28,349 @@
...
         }
 
-        if (!is_file($csvFilePath))
+        Utils::Log(LOG_INFO, "[".get_class($this)."] CSV file is [". $sCsvFilePath . "]");
+        Utils::Log(LOG_INFO, "[".get_class($this)."] Has cs header [". $this->bHasHeader . "]");
+        Utils::Log(LOG_INFO, "[".get_class($this)."] Separator used is [". $this->sCsvSeparator . "]");
+        Utils::Log(LOG_INFO, "[".get_class($this)."] Encoding used is [". $this->sCsvEncoding . "]");
+        Utils::Log(LOG_INFO, "[".get_class($this)."] Fields [". var_dump($this->aConfiguredHeaderColumns,true) . "]");
+        Utils::Log(LOG_INFO, "[".get_class($this)."] Ignored csv fields [". var_dump($this->aIgnoredCsvColumns,true) . "]");
+        Utils::Log(LOG_INFO, "[".get_class($this)."] Default values [". var_dump($this->aSynchroFieldsToDefaultValues,true) . "]");
```
[the remark this answers is not in the document]
Comment (odain-cbd) [reply core/csvcollector.class.inc.php:157]: i am not so sure. if something wrong happened this is the kind of information we will rely to troubleshoot with a client. i will check internally with support team.

[the remark this answers is not in the document]
Comment (Hipska) [reply core/csvcollector.class.inc.php:157]: Indeed, what you described are debug messages 😉

File: core/csvcollector.class.inc.php [hunk 7]
```diff
@@ -28,258 +28,347 @@
...
-	/**
-	 * Parses configured csv file to fetch data
-	 * @see Collector::Prepare()
-	 */
+    /**
+     * Parses configured csv file to fetch data
+     * @throws Exception
+     * @see Collector::Prepare()
+     */
 	public function Prepare()
```
Comment (Hipska) [inline core/csvcollector.class.inc.php:58]: Watch indentation, spaces vs tabs...

File: core/csvcollector.class.inc.php [hunk 8]
```diff
@@ -28,258 +28,347 @@
...
-        if (!is_file($csvFilePath))
+        Utils::Log(LOG_DEBUG, "[".get_class($this)."] CSV file is [". $sCsvFilePath . "]");
+        Utils::Log(LOG_DEBUG, "[".get_class($this)."] Has cs header [". $this->bHasHeader . "]");
+        Utils::Log(LOG_DEBUG, "[".get_class($this)."] Separator used is [". $this->sCsvSeparator . "]");
+        Utils::Log(LOG_DEBUG, "[".get_class($this)."] Encoding used is [". $this->sCsvEncoding . "]");
+        Utils::Log(LOG_DEBUG, "[".get_class($this)."] Fields [". var_export($this->aConfiguredHeaderColumns,true) . "]");
+        Utils::Log(LOG_DEBUG, "[".get_class($this)."] Ignored csv fields [". var_export($this->aIgnoredCsvColumns,true) . "]");
+        Utils::Log(LOG_DEBUG, "[".get_class($this)."] Default values [". var_export($this->aSynchroFieldsToDefaultValues,true) . "]");
+
+		if (!empty($this->sCsvCliCommand))
```
Comment (Hipska) [inline core/csvcollector.class.inc.php:159]: Indentation difference

File: core/csvcollector.class.inc.php [hunk 9]
```diff
@@ -28,258 +28,349 @@
...
-        while (($line = fgets($handle)) !== false) {
-            $this->csv_lines[] = rtrim(iconv($this->csv_encoding,$this->GetCharset(), $line), "\n");
+        while (($sLine = fgets($hHandle)) !== false) {
+            $this->aCsvLines[] = rtrim(iconv($this->sCsvEncoding,$this->GetCharset(), $sLine), "\n\r");
         }
 
-        fclose($handle);
-        $this->idx = 0;
+        fclose($hHandle);
 		return $bRet;
```
[the remark this answers is not in the document]
Comment (Hipska) [reply core/csvcollector.class.inc.php:191]: @odain-cbd Indentation  is not fixed yet..

**Review by Hipska — CHANGES_REQUESTED** [review]

File: core/csvcollector.class.inc.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (odain-cbd) [reply core/csvcollector.class.inc.php]: just modified it. needed some time and internal exchange to agree... :)

File: core/utils.class.inc.php [hunk 10]
```diff
@@ -371,7 +375,9 @@ static public function DoPostRequest($sUrl, $aData, $sOptionnalHeaders = null, &
 			$aHeaders = curl_getinfo( $ch );
 			if ($iErr !== 0)
 			{
-				throw new IOException("Problem opening URL: $sUrl, $sErrMsg");
+				throw new IOException("Problem opening URL: $sUrl"
+                    .PHP_EOL."    error msg: $sErrMsg"
+                    .PHP_EOL."    curl_init error code: $iErr (cf https://www.php.net/manual/en/function.curl-errno.php)");
```
[the remark this answers is not in the document]
Comment (odain-cbd) [reply core/utils.class.inc.php:380]: you are right. just fixed it

File: core/csvcollector.class.inc.php (hunk 7 above)
Comment (odain-cbd) [reply core/csvcollector.class.inc.php]: you are right. just fixed it

File: core/csvcollector.class.inc.php (hunk 8 above)
Comment (odain-cbd) [reply core/csvcollector.class.inc.php]: you are right. just fixed it

File: core/utils.class.inc.php [hunk 11]
```diff
@@ -167,9 +167,9 @@ static public function Log($iPriority, $sMessage)
 		}
 	}
 
-	static public function MockLog($MockedLogger)
+	static public function MockLog($oMockedLogger)
     {
```
Comment (Hipska) [inline core/utils.class.inc.php:171]: Mixed types of indetation, tabs vs spaces.

PS most IDE's can fix this automatically for the whole file.

**Review by Hipska — CHANGES_REQUESTED** [review]

**Pull request merged**